runtime/gc_vlhgc/SchedulingDelegate.cpp

```diff
@@ -973,65 +1302,317 @@ MM_SchedulingDelegate::calculateAutomaticGMPIntermission(MM_EnvironmentVLHGC *en
-		_edenRegionCount = desiredEdenCount;
+
+	/* Make sure that the total heap can expand enough to satisfy the desired change in eden size
+	 * If heap is fully expanded (or close to) make sure that there are enough free regions to satisfy given eden size change 
+	 */
+	intptr_t maxEdenChange = 0;
+	uintptr_t maxEdenRegionCount = _extensions->getHeap()->getHeapRegionManager()->getTableRegionCount();
+	bool edenIsVerySmall = (_edenRegionCount * 64) < maxEdenRegionCount;
+
+	if (0 == maxHeapExpansionRegions) {
+		/* 
+		 * The heap is fully expanded. Eden will be stealing free regions from the entire heap, without telling the heap to grow. 
+		 * Note: When heap is fully expanded, the eden sizing logic knows how much free memory is available in the heap, and knows to not grow too much 
+		 */
+		maxEdenChange = freeRegions;
+		_extensions->globalVLHGCStats._heapSizingData.edenRegionChange = 0;
 	} else {
-		/* there isn't enough memory left for a desired Eden. Allow Eden to shrink to free size(could be less than minimum size or 0) before
-		 * triggering an allocation failure collection (i.e. a global STW collect)
-		 */ 
-		_edenRegionCount = freeRegions;
-		Trc_MM_SchedulingDelegate_calculateEdenSize_reduceToFreeBytes(env->getLanguageVMThread(), desiredEdenCount, _edenRegionCount);
+		/* Eden will inform the total heap resizing logic, that it needs to change total heap size in order to maintain same "tenure" size */
+		maxEdenChange = maxHeapExpansionRegions;
+		intptr_t edenChangeWithSurvivorHeadroom = desiredEdenChangeSize;
+
+		/* Total heap needs to be aware that by changing eden size, the amount of survivor space might also need to change */
+		if (0 < desiredEdenChangeSize) {
+			edenChangeWithSurvivorHeadroom = desiredEdenChangeSize + (intptr_t)ceil(((double)desiredEdenChangeSize * _edenSurvivalRateCopyForward));
+		} else if ((0 > desiredEdenChangeSize) && !edenIsVerySmall) {
+			/* If eden is shrinking, only factor adjusting in survivor regions for total heap resizing when eden is not very small. 
+			 * Factoring in survivor regions when eden is tiny can lead to some innacuracies, and reduce free non-eden regions, which may impact performance 
+			 */
+			edenChangeWithSurvivorHeadroom = desiredEdenChangeSize + (intptr_t)floor(((double)desiredEdenChangeSize * _edenSurvivalRateCopyForward));
+		}
+		_extensions->globalVLHGCStats._heapSizingData.edenRegionChange = OMR_MIN(maxEdenChange, edenChangeWithSurvivorHeadroom);
 	}
+
+	desiredEdenChangeSize = OMR_MIN(maxEdenChange, desiredEdenChangeSize);
+
+	_edenRegionCount = (uintptr_t)OMR_MAX(1, ((intptr_t)_edenRegionCount + desiredEdenChangeSize));
+
 	Trc_MM_SchedulingDelegate_calculateEdenSize_Exit(env->getLanguageVMThread(), (_edenRegionCount * regionSize));
 }
 
-UDATA
+intptr_t
+MM_SchedulingDelegate::moveTowardRecommendedEdenForExpandedHeap(MM_EnvironmentVLHGC *env, double edenChangeSpeed) 
+{
+	Assert_MM_true((edenChangeSpeed <= 1) && (edenChangeSpeed >= 0));
```

Comment:
Should constants double compared with have double type too? Like `1.` and `0.`?
I guess there are more places we did not pay attention to it

Reply:
yes, there are multiple places where I omitted the `.0` (for no particularly good reason).
I'll go through the change set and add `.0` where necessary